chore(cx-api): introducing `cloud-assembly-schema` (#6724)
aws/aws-cdk: packages/@aws-cdk/cloud-assembly-schema/README.md

> ## Cloud Executable protocol

Reviewer: ```suggestion
## Cloud Assembly Protocol
```

Author: This will be `Cloud Assembly Schema`

<!--BEGIN STABILITY BANNER-->

---

![Stability: Stable](https://img.shields.io/badge/stability-Stable-success.svg?style=for-the-badge)

---
<!--END STABILITY BANNER-->

> This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.

Reviewer: This is where you should be describing the protocol, i.e., (1) what is it, (2) why does it exist, (3) what is the  protocol and (4) (if applicable) how to modify/upgrade.

Can we do the same for the `cx-api` module? I suppose this one contains the exact schema (and low level APIs), while the other contains mechanisms to read and write this schema as a file. Did I get that right?

Author: Yeap yeap, i'll do that 👍

> ## Cloud Assembly Schema

Author: Yeap, this is going to be part of the follow up PR to details the schema.

<!--BEGIN STABILITY BANNER-->

---

![Stability: Stable](https://img.shields.io/badge/stability-Stable-success.svg?style=for-the-badge)

---
<!--END STABILITY BANNER-->

This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.

> ## Cloud Assembly

Reviewer 2: In a sense this supersedes the unmaintained **cloud assembly spec** we have under [design/cloud-assembly](https://github.com/aws/aws-cdk/blob/master/design/cloud-assembly.md).

I agree with @nija-at that this requires more detail, but in order to move this forward, here's my suggestion:

1. Delete `design/cloud-assembly.md` as part of this PR (instead of actually deleting I would replace it with a file that points to this file).
2. Follow up with a subsequent PR to expand this README a bit and make sure it becomes the official spec.

> The *Cloud Assembly* is the artifact of the AWS CDK class library. It is produced as part of the

Author: I actually feel your suggestion is less clear. Why do we need to use `construct tree` and `serialization`? This README is not intended solely for developers. We will reference it from the user guide and I think we should avoid being overly technical. 

How about:

```suggestion
The *Cloud Assembly* is the output of the `synthesis` operation.
```

?

> The *Cloud Assembly* is the artifact of the AWS CDK class library. It is produced as part of the

Reviewer: Sure. what the purpose of the output is or what it contains (so as to infer the former). Otherwise, why is there even an output?

(Also, drop synthesis from backquotes)

> The *Cloud Assembly* is the artifact of the AWS CDK class library. It is produced as part of the

Author: I mention what this output contains in the following paragraph, is that sufficient?

[`cdk synth`](https://github.com/aws/aws-cdk/tree/master/packages/aws-cdk#cdk-synthesize)
command, or the [`app.synth()`](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/app.ts#L135) method invocation.

Its essentially a set of files and directories, one of which is the `manifest.json` file. It defines the set of instructions that are
needed in order to deploy the assembly directory.

> For example, when `cdk deploy` is executed, the CLI reads this file and performs its instructions:
> - Build container images.
> - Upload assets.
> - Deploy CloudFormation templates.

> Therefore, the assembly is how the CDK class library and CDK CLI (or any other consumer) should communicate. To ensure compatibility

Author: 👍

between the assembly and its consumers, we treat the manifest file as a well defined, versioned schema.

## Schema

This module contains the typescript structs that comprise the `manifest.json` file, as well as the
generated [*json-schema*](./schema/cloud-assembly.schema.json).

## Versioning

> The schema version is specified in the [`cloud-assembly.metadata.json`](./schema/cloud-assembly.schema.json) file, under the `version` property.

Author: 👍

> The schema version is specified in the [`cloud-assembly.metadata.json`](./schema/cloud-assembly.schema.json) file, under the `version` property.

Author: @nija-at Added a [module structure](https://github.com/aws/aws-cdk/blob/epolon/cli-framework-compatibility/packages/%40aws-cdk/cloud-assembly-schema/CONTRIBUTING.md#module-structure) section in the contribution guide. Let me know what you think.

> It follows semantic versioning, but with a small twist.

Author: Well, I would argue that it is semantic versioning, except that we determine what is considered a `major` bump. 

We still adhere to semantic versioning by only introducing breaking changes with `major` version bumps, except that they happen even in scenarios that normal schema versioning would consider non-breaking. 

Semantic versioning doesn't necessarily define what "breaking" means, its just a form of communication, so as long as we keep this communication honest, i.e, bumping the `major` version every time we introduce a change that we consider is breaking, we are ok. 

I actually like that we use the `semantic` terminology because it reminds people to be aware of `major` versions bumps.

When we add instructions to the assembly, they are reflected in the manifest file and the *json-schema* accordingly.
Every such instruction, is crucial for ensuring the correct deployment behavior. This means that to properly deploy a cloud assembly,
consumers must be aware of every such instruction modification.

For this reason, every change to the schema, even though it might not strictly break validation of the *json-schema* format,
is considered `major` version bump.

## How to consume

> If you'd like to consume the schema in order to do validations on `manifest.json` files, simply download it from this repo and run it against

Author: > 'download it' - how?

I can add an example `curl` command. 

> Instead of downloading it from the repo, perhaps it should be made available in npm package?

The file is available in the npm package of `cloud-assembly-schema` - but how would you access it specifically? In general, specs should actually be hosted in some domain that we own, i.e 'https://aws-cdk.io/cloud-assembly-schema/2.0.0`...
But its not entirely uncommon to simply fetch schemas from `raw.github`.  

We might setup this domain at some point.

> If you'd like to consume the schema in order to do validations on `manifest.json` files, simply download it from this repo and run it against

Reviewer: > but how would you access it specifically?

npm install ?

Wouldn't that be more accurate, in that, it would contain the latest *released* schema and not unreleased schema on master?

I am assuming here, a consumer wants to consume it programmatically for their own application. In this case, they are going to take a dependency on the cloud-assembly-schema package and that's their consumption model.

This is what we should be saying here, no?

> If you'd like to consume the schema in order to do validations on `manifest.json` files, simply download it from this repo and run it against

Author: Not necessarily. Its meant as a consumable independent json **file**,  that can be used from any language. That was the general use case i had in mind when i wrote "download". Similarly to how we download the CloudFormation spec.

> If you'd like to consume the schema in order to do validations on `manifest.json` files, simply download it from this repo and run it against

Reviewer: > Similarly to how we download the CloudFormation spec

We don't download the spec from their source code, instead from their website which is the latest published version of the spec. The equivalent for that is to obtain from this from npm repo or github repo, is it not?

Further, we download the cfn spec from the website because CF publishes it there. If CF were to publish this into a package into an artifactory like npm, we would take a dependency on this package and consume it that way, no?

> If you'd like to consume the schema in order to do validations on `manifest.json` files, simply download it from this repo and run it against

Author: > we would take a dependency on this package

If we happen to write in a language that they publish a package for, sure.

> We don't download the spec from their source code, instead from their website which always contains the latest published version of the spec

This is exactly what we should be doing at some point. 
Until then, the only way to download this specific json schema file (that is language agnostic) is by pointing to github as "our website".

> The equivalent for that is to obtain from this from npm repo or github repo.

Well, you can't really download a single file from NPM, so yeah, github repo it is.

standard *json-schema* validators.

> > For example: https://www.npmjs.com/package/jsonschema (this is how the CDK CLI does it)

Author: 👍

> Make sure you follow semantic versioning best practices, by **rejcting** to deploy cloud assemblies who's manifest version is greater than the one you expect.

Author: 👍

For example, if your consumer was built when the schema version was 1.2.0, you should reject deploying cloud assemblies with a
manifest version of 2.0.0 (or rather any higher than `1.2.0`). Note that your schema validation might actually work on those manifest files, but the deployment itself will
not work properly because it will ignore some assembly instructions. Only when you roll out an update to your consumer,

> should it accept these assemblies, by upgrading its expected schema version and taking into account the new instructions.

Author: I'll try rephrasing

> should it accept these assemblies, by upgrading its expected schema version and taking into account the new instructions.

Author: Gave it a shot. Let me know what you think.

> 

Author: 👍
